test/box-luatest/downgrade_test.lua

```diff
@@ -783,3 +724,54 @@ g.test_downgrade_user_auth_history_and_last_modified = function(cg)
         check_NO_new_auth_fields()
     end)
 end
+
+-- Check that SQL CHECK and FOREIGN KEY constraints could be downgraded.
+g.test_downgrade_sql_tuple_contraints = function(cg)
+    cg.server:exec(function()
+        local _func = box.space._func
+        local _ck = box.space._ck_constraint
+        local _fk = box.space._fk_constraint
+
+        box.execute([[CREATE TABLE t(i INT PRIMARY KEY, a INT, UNIQUE(a, i));]])
+        box.execute([[CREATE TABLE t1 (i INT PRIMARY KEY, a INT);]])
+        box.execute([[ALTER TABLE t ADD CONSTRAINT c1 CHECK (a > 10);]])
+        box.execute([[ALTER TABLE t ADD CONSTRAINT c2 CHECK (abs(a) < 100);]])
+        box.execute([[ALTER TABLE t1 ADD CONSTRAINT c3 CHECK (a * i == 99);]])
+        box.execute([[ALTER TABLE t ADD CONSTRAINT f1 ]]..
+                    [[FOREIGN KEY (a) REFERENCES t;]])
+        box.execute([[ALTER TABLE t1 ADD CONSTRAINT f2 ]]..
+                    [[FOREIGN KEY (i, a) REFERENCES t(a, i);]])
+
+        local s = box.space._space.index[2]:get{'T'}
+        local s1 = box.space._space.index[2]:get{'T1'}
+        t.assert_equals(_func:get{s.flags.constraint.C1}.body, 'a > 10')
+        t.assert_equals(_func:get{s.flags.constraint.C2}.body, 'abs(a) < 100')
+        t.assert_equals(_func:get{s1.flags.constraint.C3}.body, 'a * i == 99')
+        t.assert_equals(s.flags.foreign_key.F1.space, s.id)
+        t.assert_equals(s.flags.foreign_key.F1.field, {[1] = 0})
+        t.assert_equals(s1.flags.foreign_key.F2.space, s.id)
+        t.assert_equals(s1.flags.foreign_key.F2.field, {[0] = 1, [1] = 0})
+
+        box.schema.downgrade('2.10.4')
+
+        s = box.space._space.index[2]:get{'T'}
+        s1 = box.space._space.index[2]:get{'T1'}
+        t.assert(s.flags.constraint == nil)
+        t.assert(s.flags.foreign_key == nil)
+        t.assert(s1.flags.constraint == nil)
+        t.assert(s1.flags.foreign_key == nil)
+
+        local ck = _ck:get{s.id, 'C1'}
+        t.assert_equals(ck, {s.id, 'C1', false, 'SQL', 'a > 10', true})
+        ck = _ck:get{s.id, 'C2'}
+        t.assert_equals(ck, {s.id, 'C2', false, 'SQL', 'abs(a) < 100', true})
+        ck = _ck:get{s1.id, 'C3'}
+        t.assert_equals(ck, {s1.id, 'C3', false, 'SQL', 'a * i == 99', true})
+        local fk = _fk:get{'F1', s.id}
+        t.assert_equals(fk, {'F1', s.id, s.id, false, 'full', 'no_action',
+                             'no_action', {1}, {0}})
+        fk = _fk:get{'F2', s1.id}
+        t.assert_equals(fk, {'F2', s1.id, s.id, false, 'full', 'no_action',
+                             'no_action', {0, 1}, {1, 0}})
```

Comment:
Let's also add idempotence check as in other tests that test handlers that makes changes to system spaces.

Reply:
Fixed.